ext: lib: crypto: unify the API of CCM alogrithm

wentongwu opened the pull request:
unify the API of CCM alogrithm's implemation for TinyCrypt,
mbedTLS and cc2520 crypto device to make users easy to use.

Fixes #8339.

Signed-off-by: Wentong Wu wentong.wu@intel.com

tbursztyka reviewed (changes requested):
Why closing #14102 ? you can "git push -f" your branch if you made modifications: it will update the PR accordingly.

You still need to reply about my concern on modifying tinycrypt though.

nashif commented:
@wentongwu we can't change `ext/lib/crypto/tinycrypt/source/ccm_mode.c` in the Zephyr tree without having this go upstream first, please submit a patch to https://github.com/01org/tinycrypt and get it reviewed there first before applying that change to Zephyr.
Anything under `ext` is maintained outside of the Zephyr project usually.

wentongwu commented:
> Why closing #14102 ? you can "git push -f" your branch if you made modifications: it will update the PR accordingly.
> 
> You still need to reply about my concern on modifying tinycrypt though.

@tbursztyka Thanks. This PR is to unify the usage of tag in struct cipher_aead_pkt. Before this PR, TinyCrypt and mbedTLS have different handle for the tag, but users of the API should never make any assumptions on how they're implemented. As #8339 mentioned, I choose change TinyCrypt's API instead of alloc extra buffer. Also, I do the tests for the modified TinyCrypt, it works...

wentongwu commented:
> @wentongwu we can't change `ext/lib/crypto/tinycrypt/source/ccm_mode.c` in the Zephyr tree without having this go upstream first, please submit a patch to https://github.com/01org/tinycrypt and get it reviewed there first before applying that change to Zephyr.
> Anything under `ext` is maintained outside of the Zephyr project usually.

@nashif thanks. I will make the patch to TinyCrypt project.

wentongwu commented:
the PR(https://github.com/01org/tinycrypt/pull/37) has been created for tinycrypt.

wentongwu commented:
this PR depends on PR(https://github.com/01org/tinycrypt/pull/37)'s approve.

wentongwu commented:
@ceolin @spoorthik  please review this patch, Thanks

wentongwu commented:
@tbursztyka please review again

ceolin commented on drivers/crypto/crypto_tc_shim.c:
```diff
@@ -114,24 +114,13 @@ static int do_ccm_encrypt_mac(struct cipher_ctx *ctx,
-	 * of this and provide sufficient buffer space in output buffer to hold
-	 * both encrypted output and hash
-	 */
-	aead_op->tag = op->out_buf + op->in_len;
```
instead of changing tinycrypt api, you can just:
`memcpy(aead_op->tag, op->out_buf + op->in_len, ccm.mlen)`

ceolin replied:
this will ensure the behaviour described in our doc and do the same thing mbdetls does.

wentongwu replied:
my concern is that if tag behaved as doc described, users may just make sure the length of out_buf is op->in_len.

ceolin commented on drivers/crypto/crypto_tc_shim.c:
```diff
@@ -114,24 +114,13 @@ static int do_ccm_encrypt_mac(struct cipher_ctx *ctx,
-	 * and then increment it ccm.mlen times (while writing to it).
-	 */
-	op->out_len = op->in_len + ccm.mlen;
+	op->out_len = op->in_len;
```
doing it will cause we have another different behaviour between tc and mbedtls shims. crypto_mtls_shim.c code:
```
	/* This is equivalent to what the TinyCrypt shim does in
	 * do_ccm_encrypt_mac().
	 */
	apkt->pkt->out_len = apkt->pkt->in_len;
	apkt->pkt->out_len += ctx->mode_params.ccm_info.tag_len;
```

as you can see, mtls shim is counting tag_len in output

wentongwu replied:
ok, thanks

ceolin commented on drivers/ieee802154/ieee802154_cc2520.c:
```diff
@@ -1312,7 +1312,8 @@ static int _cc2520_crypto_ccm(struct cipher_ctx *ctx,
+				ctx->mode_params.ccm_info.tag_len);
 
 	return 0;
 }
```
these changes seem correct.

ceolin commented on ext/lib/crypto/tinycrypt/include/tinycrypt/ccm_mode.h:
```diff
@@ -202,7 +202,7 @@ int tc_ccm_generation_encryption(uint8_t *out, unsigned int olen,
+				   TCCcmMode_t c, const uint8_t *pTag);
 
 #ifdef __cplusplus
 }
```
don't like the idea of breaking tinycrypt API just to make our life easier. This pattern of adding payload + tag/IV in a contiguous buffer is common in tinycrypt.

wentongwu replied:
I change tinycrypt API just as my concerns.

ceolin reviewed (changes requested):
I believe it is better fix it without changing tinycrypt API

wentongwu commented:
> I believe it is better fix it without changing tinycrypt API

yes, I also doesn't want to change the API.  If doesn't take care the concern, I will do the shim layer's change soon. Thanks a lot for the review.

wentongwu commented:
@ceolin  @tbursztyka could you please review the new change? Thanks

tbursztyka commented on drivers/ieee802154/ieee802154_cc2520.c:
```diff
@@ -1283,7 +1283,7 @@ static int _cc2520_crypto_ccm(struct cipher_ctx *ctx,
 	}
 
-	if (apkt->tag) {
+	if (!apkt->tag) {
```
Just remove that if() altogether.

wentongwu replied:
@tbursztyka  Does I miss something for CC2520, could you please help explain more so that I can change the code better? Thanks!

tbursztyka replied:
Just remove that if () and its code block.

On ciphering phase: cc2520 is going to generate the tag, but is not going to give it back to the caller (the caller sets tag to NULL anyway).
On deciphering phase: cc2520 will need a tag depending on security level configuration

conclusion: there is no need to check tag pointer.

wentongwu replied:
@tbursztyka  Thanks!
And do we also need remove below memcpy? If remove, it doesn't behave as our doc described.

tbursztyka replied:
let me answer at the right line

wentongwu replied:
@tbursztyka  Thanks.

tbursztyka commented on drivers/ieee802154/ieee802154_cc2520.c:
```diff
@@ -1312,7 +1312,8 @@ static int _cc2520_crypto_ccm(struct cipher_ctx *ctx,
 	}
 
-	apkt->tag = apkt->pkt->out_buf + apkt->pkt->in_len;
+	memcpy(apkt->tag, apkt->pkt->out_buf + apkt->pkt->in_len,
```
just remove that line entirely.

tbursztyka replied:
cc2520 is not meant to be used as a generic ccm/uccm, but only for 15.4 security layer.

Anyway, I guess behaving properly as the crypto API defines, then le's keep this memcpy, however you'll need to test aptkt->tag before hand:
```
if (apkt->tag) {
        memcpy(...);
}
```

15.4 security layer sets it alway to NULL, so without this check you would get a fatal fault.

wentongwu replied:
Thanks

tbursztyka reviewed (changes requested):
Wrong cc2520 fix

wentongwu commented:
@ceolin could you please help take a look the new changes? Thanks a lot!

wentongwu committed:
ext: lib: crypto: unify the API of CCM alogrith

unify the API of CCM alogrithm's implemation for TinyCrypt,
mbedTLS and cc2520 crypto device to make users easy to use.

Fixes #8339.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>
